crates/vm-superio-ser/src/serial.rs

```diff
@@ -0,0 +1,208 @@
+    /// Interrupt Identification Register
+    pub interrupt_identification: u8,
+    /// Line Control Register
+    pub line_control: u8,
+    /// Line Status Register
+    pub line_status: u8,
+    /// Modem Control Register
+    pub modem_control: u8,
+    /// Modem Status Register
+    pub modem_status: u8,
+    /// Scratch Register
+    pub scratch: u8,
+    /// Transmitter Holding Buffer/Receiver Buffer
+    pub in_buffer: Vec<u8>,
+}
+
+// The following `From` implementations can be used to convert from an `SerialStateSer` to the
+// `SerialState` from the base crate and vice versa.
+impl From<&SerialStateSer> for SerialState {
+    fn from(state: &SerialStateSer) -> Self {
+        SerialState {
+            baud_divisor_low: state.baud_divisor_low,
+            baud_divisor_high: state.baud_divisor_high,
+            interrupt_enable: state.interrupt_enable,
+            interrupt_identification: state.interrupt_identification,
+            line_control: state.line_control,
+            line_status: state.line_status,
+            modem_control: state.modem_control,
+            modem_status: state.modem_status,
+            scratch: state.scratch,
+            in_buffer: state.in_buffer.clone(),
+        }
+    }
+}
+
+impl From<&SerialState> for SerialStateSer {
+    fn from(state: &SerialState) -> Self {
+        SerialStateSer {
+            baud_divisor_low: state.baud_divisor_low,
+            baud_divisor_high: state.baud_divisor_high,
+            interrupt_enable: state.interrupt_enable,
+            interrupt_identification: state.interrupt_identification,
+            line_control: state.line_control,
+            line_status: state.line_status,
+            modem_control: state.modem_control,
+            modem_status: state.modem_status,
+            scratch: state.scratch,
+            in_buffer: state.in_buffer.clone(),
+        }
+    }
+}
+
+impl Default for SerialStateSer {
+    fn default() -> Self {
+        SerialStateSer::from(&SerialState::default())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::io::sink;
+    use std::ops::Deref;
+    use vm_superio::serial::NoEvents;
+    use vm_superio::{Serial, Trigger};
+    use vmm_sys_util::eventfd::EventFd;
+
+    const RAW_INPUT_BUF: [u8; 3] = [b'a', b'b', b'c'];
+
+    struct EventFdTrigger(EventFd);
+
+    impl Trigger for EventFdTrigger {
+        type E = std::io::Error;
+
+        fn trigger(&self) -> std::io::Result<()> {
+            self.write(1)
+        }
+    }
+
+    impl Deref for EventFdTrigger {
+        type Target = EventFd;
+        fn deref(&self) -> &Self::Target {
+            &self.0
+        }
+    }
+
+    impl EventFdTrigger {
+        pub fn new(flag: i32) -> Self {
+            EventFdTrigger(EventFd::new(flag).unwrap())
+        }
+
+        pub fn try_clone(&self) -> Self {
+            EventFdTrigger((**self).try_clone().unwrap())
+        }
+    }
+
+    #[test]
+    fn test_state_ser_constructor() {
```

Comment:
This test seems superfluous.

Reply:
It is, but without it kcov doesn't consider the public fields of the struct as covered by the test_state_ser_default() test above

Reply:
oh right. kcov.. :)

Reply:
Hmm looks like this test doesn't even check some setters or some `new()` constructor, so it is indeed superfluous. I don't see how it could ever fail, so I would just remove it if it's okay for you too.

Reply:
I double checked and it turns out I was wrong. Removing the superfluous tests doesn't affect coverage. Will drop the commit.